docs/meta.rst

```diff
@@ -17,55 +17,124 @@ Taichi kernels are *lazily instantiated* and a lot of computation can happen at
 Template metaprogramming
 ------------------------
 
+Taichi tensors oftenly are used as globals. But you may use ``ti.template()``
+as type hint to pass a tensor as argument. For example:
+
 .. code-block:: python
 
     @ti.kernel
     def copy(x: ti.template(), y: ti.template()):
         for i in x:
             y[i] = x[i]
 
+    a = ti.var(ti.f32, 4)
+    b = ti.var(ti.f32, 4)
+    c = ti.var(ti.f32, 12)
+    d = ti.var(ti.f32, 12)
+    copy(a, b)
+    copy(c, d)
+
+
+As shown by the above example, template programming may enable us to reuse our
+code and improve better flexibility.
+
 
 Dimensionality-independent programming using grouped indices
 ------------------------------------------------------------
 
+However, the ``copy`` template shown above is not perfect, i.e., it can only be
+used to copy 1D tensors. What if we want to copy 2D tensors? Do we have to write
+another kernel?
+
+.. code-block:: python
+
+    @ti.kernel
+    def copy2d(x: ti.template(), y: ti.template()):
+        for i, j in x:
+            y[i, j] = x[i, j]
+
+Not necessary! Taichi provides ``ti.grouped`` syntax which enable you to get
+loop indices into a grouped vector, therefore unify different dimensionalities.
+For example:
+
 .. code-block:: python
 
     @ti.kernel
     def copy(x: ti.template(), y: ti.template()):
         for I in ti.grouped(y):
+            # I is a vector with same dimensionality with x and data type i32
+            # If y is 0D, then I = None
```

Comment:
Oh! Didn't know that:
```suggestion
            # If y is 0D, then I = ti.Vector([])
```
But actually we may preserve this so that people who know `x[None]` can quickly got our idea?

Reply:
Good point. Maybe we say `then I = ti.Vector([]), which is equivalent to "None" when used in x[I]`?

Reply:
```suggestion
            # If y is 0D, then I = ti.Vector([]), which is equivalent to `None` when used in x[I]
```